### PR TITLE
Add Terraform deployment engine for cloud resources

### PR DESCRIPTION
**Pull request by jimright**
Added option for Terraform deployment of the cloud resources at the infrastructure and platform run-levels.
Features:
* A parameter `infra_deployment_engine` used to determine the choice of deployment engine (pure-Ansible or Terraform) for cloud resources.
* Terraform deployment is driven by Ansible, which creates necessary .tf resource files from Jinja templates and runs terraform apply\destroy. 
* Currently only supports the AWS `infra_type`:
  * Creates network, storage and compute resources at the infra level
  * Creates the AWS IAM policies and roles at the plat level
* Support for `local` and `remote_s3` (S3 bucket for storage; DynamoDB for state locking) backend storage of the Terraform state.
* A timestamped artefact directory with a copy of the Terraform files created at each execution of cloudera-deploy.

Draft PR has a `debug_terraform` bool variable. This puts pause points before the main terraform steps allowing for review. These will be cleaned before the final PR.

### CODE REVIEW

**Comment by jimright** [comment]
Have rebased the code in this PR following the 1.5.1 release and retested that L0 and L1 deployment and teardown with Terraform work ok.

File: roles/platform/tasks/initialize_setup_aws.yml [hunk 1]
```diff
@@ -78,6 +78,7 @@
 - name: Handle AWS Public and Private VPC Subnets if not defined
   when: not plat__aws_public_subnet_ids or not plat__aws_private_subnet_ids
   block:
+    # Search using a wildcard on the subnet Name tag {{ plat__namespace }}*
```
Comment (Chaffelson) [inline roles/platform/tasks/initialize_setup_aws.yml:81]: I'm confused by this comment

File: roles/platform/tasks/setup_aws_terraform_authz.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    src: "{{ plat__terraform_template_dir }}/plat/"
+    dest: "{{ plat__terraform_artefact_dir }}/plat"
+
+- name: Prompt to debug Terraform deployment 
+  ansible.builtin.pause:
+    prompt: "Ready to do Terraform apply on role files in {{ plat__terraform_workspace_dir }}/workspace/plat"
+  when: debug_terraform | default(false) | bool
+
+- name: Applying Terraform
+  community.general.terraform:
```
Comment (Chaffelson) [inline roles/platform/tasks/setup_aws_terraform_authz.yml:39]: logging?

File: roles/platform/tasks/teardown_aws_terraform_authz.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  # when: not workdir.stat.exists # TODO: This won't override if any changes to .tf are made. Good or Bad?
+
+- name: Prompt to debug Terraform deployment 
+  ansible.builtin.pause:
+    prompt: "Ready to do Terraform destroy on role files in {{ plat__terraform_workspace_dir }}/workspace/plat"
+  when: debug_terraform | default(false) | bool
+
+- name: Destroy Terraform authz resources
+  community.general.terraform:
+    project_path: "{{ plat__terraform_workspace_dir }}/workspace/plat"
```
Comment (Chaffelson) [inline roles/platform/tasks/teardown_aws_terraform_authz.yml:41]: logz

File: roles/platform/tasks/teardown_aws_terraform_authz.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  ansible.builtin.file:
+    path: "{{ plat__terraform_template_dir }}/plat/"
+    state: absent
+
+- name: Remove the Terraform workspace directory
+  ansible.builtin.file:
+    path: "{{ plat__terraform_workspace_dir }}/workspace/plat"
+    state: absent
+    
+# TODO: Figure out how to handle below conditional with Terraform
```
Comment (Chaffelson) [inline roles/platform/tasks/teardown_aws_terraform_authz.yml:64]: TBD?

File: roles/common/defaults/main.yml [hunk 5]
```diff
@@ -57,6 +57,7 @@ common__ngw_suffix:                       "{{ globals.labels.nat_gateway | defau
 common__unique_storage_name_suffix:       "{{ globals.storage.name | default((common__region + common__aws_profile) if 'aws' in common__infra_type else common__region) }}"
 
 # Infra
+common__infra_deployment_engine:          "{{ globals.infra_deployment_engine | default() }}"
```
[the remark this answers is not in the document]
Comment (jimright) [reply roles/common/defaults/main.yml]: Good spot, changed that in the latest update of the code.

File: roles/common/defaults/main.yml [hunk 6]
```diff
@@ -66,6 +67,19 @@ common__public_key_text:                  "{{ globals.ssh.public_key_text | defa
 common__region:                           "{{ globals.region | default(common__region_default[common__infra_type]) }}"
 common__storage_name:                     "{{ infra.storage.name | default([common__namespace, common__unique_storage_name_suffix[::2] | replace('-','')] | join('-')) }}"
 
+# Terraform
+# The processed Jinja template files for Terraform are placed in common__terraform_template_dir
+common__terraform_template_dir:            "{{ globals.terraform_template_dir | default( ( playbook_dir , 'terraform_processed_template_code') | path_join ) }}"
```
[the remark this answers is not in the document]
Comment (jimright) [reply roles/common/defaults/main.yml]: Great idea. I've added the `terraform_base_dir` variable and all other Terraform directories are derived from this.

File: roles/infrastructure/tasks/setup_terraform.yml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  ansible.builtin.copy:
+    src: "{{ infra__terraform_template_dir }}/infra/"
+    dest: "{{ infra__terraform_artefact_dir }}/infra"
+
+- name: Prompt to debug Terraform deployment 
+  ansible.builtin.pause:
+    prompt: "Ready to do Terraform apply on infra files in {{ infra__terraform_workspace_dir }}/workspace/infra"
+  when: debug_terraform | default(false) | bool
+
+- name: Applying Terraform
```
[the remark this answers is not in the document]
Comment (jimright) [reply roles/infrastructure/tasks/setup_terraform.yml:38]: I'm not sure about how to increase the verbosity by default. Will have to look into this a bit more.

File: roles/infrastructure/tasks/setup_terraform.yml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    path: "{{ infra__terraform_template_dir }}/infra"
+    state: absent
+
+- name: Remove the Terraform workspace directory when using remote state
+  ansible.builtin.file:
+    path: "{{ infra__terraform_workspace_dir }}/workspace/infra"
+    state: absent
+  when: infra__terraform_state_storage in ['remote_s3']
+
+# If created Utility Instance via Terraform then
```
[the remark this answers is not in the document]
Comment (jimright) [reply roles/infrastructure/tasks/setup_terraform.yml:64]: The Terraform and Ansible set of tasks are a bit different for this. Ansible creates the utility instance and adds it to the inventory; Terraform discovers the utility instance and adds to inventory. So I think it best to keep these tasks separate.

File: roles/infrastructure/tasks/validate_aws_terraform.yml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+- name: Print AWS Profile to debug
+  ansible.builtin.command: echo AWS_PROFILE is $AWS_PROFILE
+
+- name: Confirm that required Terraform variables are defined
+  block:
+  - name: Check infra__terraform_template_dir
+    ansible.builtin.assert:
+      that: 
+        - "infra__terraform_template_dir is defined"
```
[the remark this answers is not in the document]
Comment (jimright) [reply roles/infrastructure/tasks/validate_aws_terraform.yml:38]: Good spot, changed that in the latest update of the code.

File: roles/platform/tasks/initialize_aws_terraform.yml [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    state: directory
+
+# Create and process the AWS Policy documents
+- name: Create a temporary directory for policy documents
+  ansible.builtin.tempfile:
+    prefix: "aws-policy-"
+    state: directory
+  register: __aws_policy_tmpdir
+
+- name: Download AWS default policy documents
```
[the remark this answers is not in the document]
Comment (jimright) [reply roles/platform/tasks/initialize_aws_terraform.yml]: Agree. In my updated changes, I've moved the download and process of AWS policy documents to a single task file, `aws_policy_download.yml`. 
I changed the below task files to then include this task file:
* setup_aws_autz.yml
* initialize_aws_terraform.yml, 
* initialize_teardown_aws_terraform.yml

File: roles/platform/tasks/initialize_setup_aws.yml (hunk 1 above)
Comment (jimright) [reply roles/platform/tasks/initialize_setup_aws.yml]: Sorry, a hold-over from an older version of my code. Removed the comment now.

File: roles/platform/tasks/teardown_aws_terraform_authz.yml (hunk 4 above)
Comment (jimright) [reply roles/platform/tasks/teardown_aws_terraform_authz.yml]: Yes I need to figure out how to skip teardown of Terraform-created resources if some flags (e.g. env.teardown.delete_cross_account: False) are specified. I've removed the TODO from the code and will create an issue for these tasks.

**Comment by jimright** [comment]
Hey @Chaffelson 

Thanks for looking through this. 
I've changed and updated cloudera.exe based on your comments. 
See above replies to your comments for specific details of changes.

Together with the updates to cloudera-deploy, this should be ready to retest now.

**Review by Chaffelson — APPROVED** [review]

**Comment by jimright** [comment]
@Chaffelson @wmudge,

Have updated the cloudera.exe Terraform PR following the release of the AWS provider. 
This new provider version had some breaking changes and deprecations - listed [here](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade) - so I've handled those and tested to confirm everything still works.

Thanks,
Jim

File: roles/infrastructure/tasks/setup_terraform.yml [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    path: "{{ infra__terraform_template_dir }}/infra"
+    state: absent
+
+- name: Remove the Terraform workspace directory when using remote state
+  ansible.builtin.file:
+    path: "{{ infra__terraform_workspace_dir }}/infra"
+    state: absent
+  when: infra__terraform_state_storage in ['remote_s3']
+
+# Get information about Dynamic Inventory VMs if created via Terraform
```
Comment (wmudge) [inline roles/infrastructure/tasks/setup_terraform.yml:53]: Could this have been done using output variables from TF?

File: roles/infrastructure/tasks/setup_terraform.yml [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  loop_control:
+    loop_var: __infra_di_item
+
+# If created Utility Instance via Terraform then
+# need to get it's info and add to an Ansible host group
+- name: Add Utility Instance to host group
+  when: infra__create_utility_service
+  block:
+    - name: Discover the Utility Instance details
+      community.aws.ec2_instance_info:
```
Comment (wmudge) [inline roles/infrastructure/tasks/setup_terraform.yml:79]: Use output variables in TF?

File: roles/infrastructure/tasks/teardown_terraform.yml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: Check if Terraform workspace directory exists
+  ansible.builtin.stat: 
+    path: "{{ infra__terraform_workspace_dir }}/infra"
+  register: workdir
```
Comment (wmudge) [inline roles/infrastructure/tasks/teardown_terraform.yml:20]: Where is this registered variable used?

File: roles/infrastructure/tasks/main.yml [hunk 14]
```diff
@@ -15,5 +15,5 @@
 # limitations under the License.
 
 - include_tasks: validate.yml
-- include_tasks: initialize_setup.yml
-- include_tasks: setup.yml
+# - include_tasks: initialize_setup.yml
```
Comment (wmudge) [inline roles/infrastructure/tasks/main.yml:18]: Why is this commented out?

**Review by wmudge — CHANGES_REQUESTED** [review]

File: roles/infrastructure/tasks/main.yml (hunk 14 above)
Comment (jimright) [reply roles/infrastructure/tasks/main.yml]: Apologies, this is a hangover from when I first started this PR and trying to understand cloudera.exe. Uncommented these lines in [617d68c](https://github.com/cloudera-labs/cloudera.exe/pull/56/commits/617d68c30954838b9e3c921de9e84d92dc594445)

File: roles/infrastructure/tasks/setup_terraform.yml (hunk 11 above)
Comment (jimright) [reply roles/infrastructure/tasks/setup_terraform.yml:53]: Yes, I will explore if using tf output is possible and submit a separate PR for this.

File: roles/infrastructure/tasks/setup_terraform.yml (hunk 12 above)
Comment (jimright) [reply roles/infrastructure/tasks/setup_terraform.yml:79]: As above, I will test this and submit a separate PR if it looks good.

File: roles/infrastructure/tasks/teardown_terraform.yml (hunk 13 above)
Comment (jimright) [reply roles/infrastructure/tasks/teardown_terraform.yml]: Good spot, this is no longer used. There are a few similar instances of this in the code that I have cleaned as part of 
287c30d.

**Comment by jimright** [comment]
@wmudge. 

Thanks for looking through the PR. I have updated the code to address most of the issues you spotted. Also retested the AWS Terraform deployment and teardown to make sure that still works.

For things like using terraform output instead of Ansible-based discovery, I will test this separately and create a PR if all goes well.

Jim

**Review by wmudge — APPROVED** [review]

**Pull request merged**